### PR TITLE
Make find-file-at-point support <nixpkgs>

### LINKED ISSUE #58: Support find-file-at-point or something like it, to open a file referenced by your nix expression
It would be super neato to support the following use cases:

(1) relative paths, pretty easy since it is relative to the current file!

```nix
{
  imports = [
    ./macs/nodes/mac5.nix
  ];
}
```

having the cursor at the beginning of `./macs/nodes/mac5.nix`, being able to do C-c C-o or something to open the file.

(2) paths based on NIX_PATH

 a bit trickier, but would be pretty neat:

```nix
{ pkgs ? import <nixpkgs> }:
null
```

being on `<nixpkgs>` and have it open default.nix of the NIX_PATH's nixpkgs. This can be done with nix-instantiate:

```
grahamc@Morbo> nix-instantiate  --eval -E '<nixpkgs>'
/nix/var/nix/profiles/per-user/root/channels/nixos
```

of course, this should also support sub-files using the same method:

```
grahamc@Morbo> nix-instantiate  --eval -E '<nixpkgs/nixos/lib/make-channel.nix>'
/nix/var/nix/profiles/per-user/root/channels/nixos/nixos/lib/make-channel.nix
```

Ideally version 2 would also support direnv altering the NIX_PATH.

### PR DESCRIPTION
**Pull request by jabranham**
Closes #58 

I think this also respects direnv but I don't use it so it would be nice if someone could check.

### CODE REVIEW

File: nix-mode.el [hunk 1]
```diff
@@ -520,6 +520,22 @@ STRING-TYPE type of string based off of Emacs syntax table types"
     (when (> end-of-indentation (point)) (goto-char end-of-indentation)))
   )
 
+;; ffap support
+(defun nix-mode-ffap-nixpkgs-path (str)
+  "Support `ffap' for <nixpkgs> declarations.
+If STR contains nixpkgs, call nix-instantiate to find the
+location of STR."
+  ;; The < > around <nixpkgs> don't get passed to us
+  (when (string-match "nixpkgs" str)
```
Comment (grahamc) [inline nix-mode.el]: Hmmm... can this be altered to generally match any path within `<...>`s? It won't always be `<nixpkgs...>` but instead could be almost anything.

**Comment by grahamc** [comment]
After applying this patch locally, I get:

```
In nix-mode-ffap-nixpkgs-path:
nix-mode.el:439:26:Warning: reference to free variable
    `nix-instantiate-executable'
nix-mode.el:481:1:Warning: unescaped character literals `?(', `?)', `?[', `?]'
    detected!
Compiling /nix/store/ljaxyxk4vf674w1kh17m878al5cwjvx8-emacs-nix-mode-20180629.1501/share/emacs/site-lisp/elpa/nix-mode-20180629.1501/nix-prettify-mode.el...
```

When I call `find-file-at-point` on `<nixpkgs/nixos/release.nix>` (or directly inside the `<`) it doesn't seem to locate the file.

File: nix-mode.el (hunk 1 above)
Comment (jabranham) [reply nix-mode.el]: The `<` and `>` characters don't get passed through to us, unfortunately. So we can't tell if we're inside <> or not.

**Comment by jabranham** [comment]
I'm not sure why you'd get a warning about nix-instantiate-executable not being defined yet. It's in nix.el, which is required by nix-mode.el 

It won't work if point is on the `<`. So if `|` is point, `<|nixpkgs/nixos/release.nix>` M-x ffap should result in something like `/nix/var/nix/profiles/per-user/root/channels/nixos/nixos/release.nix`. Did `(nix-mode . nix-mode-ffap-nixpkgs-path)` get added to `ffap-alist` for you?

File: nix-mode.el [hunk 2]
```diff
@@ -520,6 +520,22 @@ STRING-TYPE type of string based off of Emacs syntax table types"
...
+    (let ((file (shell-command-to-string
+                 (concat nix-instantiate-executable " --eval -E '<" str ">'"))))
+      ;; Remove trailing newline
+      (replace-regexp-in-string "\n$" "" file))))
```
Comment (matthewbauer) [inline nix-mode.el]: To avoid the regex you can do ```(substring 0 (- (length file) 1) file)

Comment (matthewbauer) [reply nix-mode.el]: Although I guess that kind of breaks when we don't have a trailing new line.

File: nix-mode.el (hunk 1 above)
Comment (matthewbauer) [reply nix-mode.el]: You can keep the brackets with:

```
(push '(nix-mode "--:\\\\${}<>+@-Z_[:alpha:]~*?" "@" "@;.,!:") ffap-string-at-point-mode-alist)
```

**Comment by matthewbauer** [comment]
Merged in 9d4eb7d199bcefb602db1a8332761f595410899c

**Pull request closed without merging**